### PR TITLE
Fixed out of memory handling in write-behind process

### PR DESCRIPTION
**Pull request by ahmetmircik**
Now it does not swallow OutOfMemoryError. Plus, changed log level of store failure from warning to error.

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/DefaultWriteBehindProcessor.java [hunk 1]
```diff
@@ -372,6 +372,9 @@ public void callAfterStoreListeners(Collection<DelayedEntry> entries) {
             try {
                 result = task.run();
             } catch (Throwable t) {
+                if (t instanceof Error) {
```
Comment (gurbuzali) [inline hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/DefaultWriteBehindProcessor.java]: why don't you just catch `Exception` not `Throwable` ?

Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/DefaultWriteBehindProcessor.java]: fixed.

**Comment by gurbuzali** [comment]
:+1:

**Comment by ahmetmircik** [comment]
[Running com.hazelcast.client.mapreduce.aggregation.AvgAggregationTest] (https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-pr-builder/14115/console)
15:48:49 Build timed out (after 150 minutes). Marking the build as aborted.

**Comment by ahmetmircik** [comment]
verify

File: hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/DefaultWriteBehindProcessor.java [hunk 2]
```diff
@@ -382,11 +382,14 @@ public void callAfterStoreListeners(Collection<DelayedEntry> entries) {
...
-                    result ? "succeeded." : "failed too.");
-            logger.warning(msg, throwable);
             if (!result) {
-                return task.failureList();
+                // List of entries which can not be stored for this round. We will readd these entries
+                // in front of the relevant partition-write-behind-queues and will indefinitely retry to
+                // store them.
+                List failureList = task.failureList();
+                logger.severe("Number of entries which could not be stored is = [" + failureList.size() + "]"
+                        + ", Hazelcast will indefinitely retry to store them", exception);
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/DefaultWriteBehindProcessor.java]: Do we need to add the word 'indefinitely' to the logging?

**Comment by pveentjer** [comment]
Perhaps I'm missing something, but why does the system retry of an OOME happens? Afaik the HZ instance normally is shutting down. Can you provide some background information for the change?

**Comment by ahmetmircik** [comment]
You are right @pveentjer, the system should not retry an OOME. This PR is fixing that problem.

**Pull request merged**